Source/UI/NewUI/PaymentForm/Factory/PaymentFormFactory.swift

```diff
@@ -9,30 +9,28 @@ public struct PaymentFormFactory {
   internal static var sessionCorrelationID = ""
 
   public static func buildViewController(configuration: PaymentFormConfiguration,
-                                         style: PaymentStyle) -> UIViewController {
+                                         style: PaymentStyle,
+                                         cardTokenResultClousure: @escaping (Result<TokenDetails, TokenisationError.TokenRequest>) -> Void) -> UIViewController {
```

Comment:
Any specific reason why you chose a completion handler over a delegate for this callback?

Reply:
If you see any pros for this solution, please suggest code.

Reply:
I think code reviews are a good place for having discussions around implementation and details. I'm trying to understand why you chose to go for a completion handler in the factory versus injecting a delegate.

Personally I find delegates safer for retain cycles as it gives us the control over weakening the callback reference and helps provide a safer solution for consumer of SDK.

Reply:
My high level understanding was, since we are using callbacks everywhere in the code we are preferring callbacks and not using delegate. More of the design choice between callback or delegate. @ehab-al-cko Please correct me if I am wrong on this.

Reply:
I don't think that assertion should apply here as this is the only place the user gets to catch the output of the journey. All the others are internal, between the VC and VM.

Consumer will say

> I want to start a tokenisation request with this input and on output I want this to happen

Anything else is hidden from consumer (dev team). And completion handlers whilst they read easier at declaration site are more prone for implementation errors. And whilst we are not responsible for our consumers technical expertise, making our solution work for more people at no cost (delegate pattern is common in iOS) might be a nice small detail of our product.

Reply:
Apple and Swift are now moving away from delegates towards callbacks and eventually `async`/`await`s. Delegate methods are tried and true but if we want to follow examples from newer Apple APIs, such as those in SwiftUI, then this should be a callback.

Reply:
> Apple and Swift are now moving away from delegates towards callbacks

I think that is a pretty bold statement. SwiftUI is a different programming paradigm, and async/await tries to serialise async logic for reader.

***
> Delegate methods are tried and true but if we want to follow examples from newer Apple APIs, such as those in SwiftUI, then this should be a callback.

Given SwiftUI is declarative UI, the whole paradigm is different so that does not mean delegates are discouraged or becoming obsolete. 
Delegates, completion handlers, bindings are all ways for achieving the same task which is to communicate from child to parent. SwiftUI being declarative works naturally with bindings. That is not relevant therefore to our assumptions on delegates vs completion handlers, unless if I missed or misunderstood something?